src/main/scala/edu/uci/eecs/spectralLDA/datamoments/DataCumulantSketch.scala

```diff
@@ -143,45 +182,122 @@ object DataCumulantSketch {
+    *                  \frac{\alpha_0}{\alpha_0+2}\left(E[x_1\otimes x_2\otimes M1]
+    *                                       +E[x_1\otimes M1\otimes x_2]
+    *                                       +E[M1\otimes x_1\otimes x_2]\right)(W^T,W^T,W^T)$
+    *         Refer to Eq (22) in [Wang2015]
+    *
+    * REFERENCES
+    * [Wang2015] Wang Y et al, Fast and Guaranteed Tensor Decomposition via Sketching, 2015,
+    *            http://arxiv.org/abs/1506.04448
+    *
+    */
+  private def update_thirdOrderMoments(alpha0: Double,
+                                       W: DenseMatrix[Double],
+                                       q: DenseVector[Double],
+                                       n: SparseVector[Double],
+                                       len: Double,
+                                       sketcher: TensorSketcher[Double, Double])
+        : DenseMatrix[Complex] = {
+    /* ------------------------------------- */
+
+    // $p=W^T n$, where n is the original word count vector
+    val p: DenseVector[Double] = W.t * n
+
+    // fft of sketch_p, $p=W^T n$, where $n$ is the original word count vector
+    val fft_sketch_p: Seq[DenseMatrix[Complex]] = (0 until 3)
+      .map { sketcher.sketch(p, _) }
+      .map { A: DenseMatrix[Double] => fourierTr(A(*, ::)) }
+
+    // fft of sketch_q, $q=W^T M1$
+    val fft_sketch_q: Seq[DenseMatrix[Complex]] = (0 until 3)
+      .map { sketcher.sketch(q, _) }
+      .map { A: DenseMatrix[Double] => fourierTr(A(*, ::)) }
+
+
+    /* ------------------------------------- */
+
+    // sketch of $p^{\otimes 3}$
+    val fft_sketch_p_otimes_3 = fft_sketch_p(0) :* fft_sketch_p(1) :* fft_sketch_p(2)
+
+    // sketch of $p\otimes p\otimes q$
+    val fft_sketch_p_p_q = fft_sketch_p(0) :* fft_sketch_p(1) :* fft_sketch_q(2)
+
+    // sketch of $p\otimes q\otimes p$
+    val fft_sketch_p_q_p = fft_sketch_p(0) :* fft_sketch_q(1) :* fft_sketch_p(2)
+
+    // sketch of $q\otimes p\otimes p$
+    val fft_sketch_q_p_p = fft_sketch_q(0) :* fft_sketch_p(1) :* fft_sketch_p(2)
+
+    /* ------------------------------------- */
+
+    // fft of sketch of $\sum_{i=1}^V -n_i\left(w_i\otimes w_i\otimes p+w_i\otimes p\otimes w_i
+    //                                           +p\otimes w_i\otimes w_i\right)
+    //                +\sum_{i=1}^V 2n_i w_i^{\otimes 3}$
+    // ref: Eq (25) in [Wang2015]
+    val fft_sum1: DenseMatrix[Complex] = fft_sketch_q(0) * Complex(0, 0)
+
+    // fft of sketch of $\sum_{i=1}^V -n_i\left(w_i\otimes w_i\otimes q
+    //                                          +w_i\otimes q\otimes w_i
+    //                                          +q\otimes w_i\otimes w_i\right)$
+    // ref: Eq (26) in [Wang2015]
+    val fft_sum2: DenseMatrix[Complex] = fft_sketch_q(0) * Complex(0, 0)
+
+    for ((wc_index, wc_value) <- n.activeIterator) {
+      val fft_sketch_w_i: Seq[DenseMatrix[Complex]] = (0 until 3)
+        .map { sketcher.sketch(W(wc_index, ::).t, _) }
```

Comment:
For sure code-wise, it's more efficient to combine any two consecutive maps.

I chose to write them in two for cleanness, and aesthetics. I found it'd be quicker and clearer to understand what's going on for a reader. I have a belief that visually clean, concise and easy code saves time overall (running time, maintenance time) -- but it's a personal view.